File: packages/android/app/src/main/java/app/keyweb/sharing/BackupFile.kt
package app.keyweb.sharing

import app.keyweb.vault.HLC_ZERO
import app.keyweb.vault.Reg
import app.keyweb.vault.VaultState
import app.keyweb.vault.boundDatasets
import app.keyweb.vault.extractDataset
import app.keyweb.vault.withoutDatasetItems
import com.keyneom.synckit.crypto.SyncKitJson
import com.keyneom.synckit.sharing.CreateSharedBackupEnvelopeInput
import com.keyneom.synckit.sharing.ProtectedSharingIdentityV1
import com.keyneom.synckit.sharing.SharedBackupCodec
import com.keyneom.synckit.sharing.SharedBackupEnvelopeV1
import com.keyneom.synckit.sharing.SharedBackupParticipantInput
import com.keyneom.synckit.sharing.SharingCrypto
import com.keyneom.synckit.sharing.SharingIdentity
import com.keyneom.synckit.sharing.SharingRole
import com.keyneom.synckit.sharing.VerifySharedBackupOptions
import java.util.UUID
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

/**
 * An encrypted backup file, for the day Google is the thing that is down.
 *
 * The counterpart of the web's `backupFile.ts`, byte for byte in format, so a
 * file saved on either opens on the other. It is the vault sealed exactly as a
 * keyring file is — encrypted once, its key wrapped to you — with the recovery
 * lock carried beside it. The printed code opens the lock, the lock makes you
 * *you*, and you open the vault: no Google, no passkey, nothing but the file
 * and the code.
 */

const val BACKUP_FILE_FORMAT = "keyweb-backup"

@Serializable
data class BackupFileV1(
    val format: String,
    val version: Int,
    val savedAt: String,
    val recoveryLock: ProtectedSharingIdentityV1,
    val envelope: SharedBackupEnvelopeV1,
)

/** What opening a file gives back: the vault, and who you are. */
class OpenedBackupFile(
    val state: VaultState,
    val identity: SharingIdentity,
    val recoveryLock: ProtectedSharingIdentityV1,
)

/** The code does not open this file. Said plainly, not as a crypto error. */
class WrongBackupCode : Exception("That recovery code doesn't open this backup file.")

class NotABackupFile(message: String) : Exception(message)

private val vaultJson = Json { ignoreUnknownKeys = true }

private object BackupCodec : SharedBackupCodec<VaultState> {
    override fun serialize(value: VaultState): JsonElement =
        vaultJson.encodeToJsonElement(VaultState.serializer(), value)

    override fun parse(value: JsonElement): VaultState =
        vaultJson.decodeFromJsonElement(VaultState.serializer(), value)
}

/**
 * Seal the vault into a file the printed code opens.
 *
 * Refuses a lock on anyone but [identity]: a file whose lock opens somebody
 * else would open nothing, and look as though it would.
 */
fun sealBackupFile(
    state: VaultState,
    identity: SharingIdentity,
    recoveryLock: ProtectedSharingIdentityV1,
    now: String = java.time.Instant.now().toString(),
): String {
    require(recoveryLock.publicKey.keyId == identity.publicKey.keyId) {
        "That recovery lock is not yours."
    }
    val envelope = SharingCrypto.createSharedBackupEnvelopeV1(
        state,
        BackupCodec,
        identity,
        CreateSharedBackupEnvelopeInput(
            appId = "keyweb-backup",
            backupId = "backup-${UUID.randomUUID()}",
            participants = listOf(
                SharedBackupParticipantInput(publicKey = identity.publicKey, role = SharingRole.OWNER),
            ),
            createdAt = now,
        ),
    )
    val file = BackupFileV1(
        format = BACKUP_FILE_FORMAT,
        version = 1,
        savedAt = now,
        recoveryLock = recoveryLock,
        envelope = envelope,
    )
    return SyncKitJson.instance.encodeToString(BackupFileV1.serializer(), file)
}

/** Open a backup file with the printed code alone. */
fun openBackupFile(text: String, code: ByteArray): OpenedBackupFile {
    val file = try {
        SyncKitJson.instance.decodeFromString(BackupFileV1.serializer(), text)
    } catch (cause: Exception) {
        throw NotABackupFile("That isn't a Keyweb backup file.")
    }
    if (file.format != BACKUP_FILE_FORMAT || file.version != 1) {
        throw NotABackupFile("That isn't a Keyweb backup file.")
    }
    val identity = try {
        unlockRecoveryLock(file.recoveryLock, code)
    } catch (cause: Exception) {
        throw WrongBackupCode()
    }
    val state = SharingCrypto.decryptSharedBackupEnvelopeV1(
        file.envelope,
        BackupCodec,
        identity,
        VerifySharedBackupOptions(trustedOwnerKeyId = identity.publicKey.keyId),
    )
    return OpenedBackupFile(restorable(state), identity, file.recoveryLock)
}

/**
 * A backed-up vault, ready to go into one that may have no Drive to reach.
 *
 * Which file each keyring lives in is dropped: the restoring vault already
 * knows where its own keyrings live, and a binding from the file could point
 * at a file this account no longer has. Dropped as the oldest possible value
 * rather than an empty newer one, so a binding the vault holds always wins.
 */
private fun restorable(state: VaultState): VaultState =
    state.copy(
        keyrings = state.keyrings.mapValues { (_, keyring) -> keyring.copy(dataset = Reg("", HLC_ZERO)) },
    )

/**
 * Where each part of a restored vault goes, by the vault it is going into.
 *
 * A keyring kept in a file of its own gets its passwords in that file's
 * document, not the vault's; otherwise they would be published into the
 * index beside keyrings they were never meant to travel with. Returned as
 * document id to state, with the vault itself under `""`.
 */
fun restorePlan(current: VaultState, restored: VaultState): Map<String, VaultState> {
    val plan = mutableMapOf<String, VaultState>()
    var rest = restored
    for (binding in boundDatasets(current)) {
        val slice = extractDataset(rest, binding.keyringId)
        if (slice.items.isNotEmpty()) plan[binding.datasetId] = slice
        rest = withoutDatasetItems(rest, binding.keyringId)
    }
    plan[""] = rest
    return plan
}

File: packages/web/src/vault/backupFile.ts
import {
  createSharedBackupEnvelopeV1,
  decryptSharedBackupEnvelopeV1,
  type WebCryptoSharingIdentity,
} from "@keyneom/sync-kit/sharing/web-crypto";
import type { SharedBackupEnvelopeV1 } from "@keyneom/sync-kit/sharing";
import type {
  ProtectedSharingIdentityStore,
  ProtectedSharingIdentityV1,
} from "@keyneom/sync-kit/sharing/web-passkey";
import {
  boundDatasets,
  extractDataset,
  keyringsOf,
  type VaultState,
  withoutDatasetItems,
} from "@keyweb/vault-core";
import { KEYWEB_RECOVERY_APP_ID, unlockRecoveryIdentity } from "./sharing/identity";

/**
 * An encrypted backup you keep yourself, and open with your printed code.
 *
 * ## What it is for
 *
 * Everything else Keyweb keeps lives in your Google account: the keyring files,
 * the index, and the lock your recovery code opens. That covers losing a
 * passkey. It does not cover losing Google — an outage, a locked account —
 * because then there is nothing for the code to open. This file is the one
 * thing that survives that: it holds your passwords *and* the lock, so the file
 * and the code on paper are enough, with no network at all.
 *
 * ## How it is sealed
 *
 * The same way every keyring file is: the vault encrypted once, its key granted
 * to your identity through sync-kit's envelope. The recovery lock rides beside
 * it — your identity wrapped by your code — so the code gives back you, and you
 * open the envelope. Neither half is readable without the code.
 *
 * It holds the vault's CRDT state rather than the plain export, because this is
 * a backup to restore rather than a file for another app: restoring is a merge,
 * which keeps history, files and timestamps and cannot lose anything newer that
 * is already on the device.
 */
export const BACKUP_FILE_FORMAT = "keyweb-backup";

export type BackupFileV1 = {
  format: typeof BACKUP_FILE_FORMAT;
  version: 1;
  savedAt: string;
  /** Your identity, wrapped by your printed recovery code. */
  recoveryLock: ProtectedSharingIdentityV1;
  /** Your vault, encrypted once, its key granted to that identity. */
  envelope: SharedBackupEnvelopeV1;
};

const codec = {
  serialize: (value: VaultState) => value,
  parse: (value: unknown) => value as VaultState,
};

/**
 * Seal a vault into a backup file.
 *
 * `recoveryLock` has to be the lock for *this* identity — the one the phone
 * writes to app-data — or the file would open for nobody. Checked by keyId
 * rather than trusted.
 */
export async function sealBackupFile(
  state: VaultState,
  identity: WebCryptoSharingIdentity,
  recoveryLock: ProtectedSharingIdentityV1,
  now: Date = new Date(),
): Promise<BackupFileV1> {
  if (recoveryLock.publicKey.keyId !== identity.publicKey.keyId) {
    throw new Error("That recovery lock belongs to a different identity.");
  }
  const envelope = await createSharedBackupEnvelopeV1(state, codec, identity, {
    appId: "keyweb-backup",
    backupId: `backup-${crypto.randomUUID()}`,
    participants: [{ publicKey: identity.publicKey, role: "owner" }],
    createdAt: now.toISOString(),
  });
  return {
    format: BACKUP_FILE_FORMAT,
    version: 1,
    savedAt: now.toISOString(),
    recoveryLock,
    envelope,
  };
}

/**
 * Open a backup file with the printed code — no network, no passkey.
 *
 * Throws on a wrong code, a file that is not a Keyweb backup, or one whose
 * envelope was not signed by the identity its own lock holds.
 */
export async function openBackupFile(input: unknown, code: Uint8Array): Promise<VaultState> {
  return (await openBackupFileAsYou(input, code)).state;
}

/**
 * Open a backup file, and give back who you are along with what was in it.
 *
 * The lock inside the file is the one beside your passkey in the Google
 * account, so opening it makes you *you* — the same identity every one of your
 * keyring files is wrapped to. A browser restoring from the file keeps both:
 * the identity for the session, so your files open once Google is back, and
 * the lock, so the code opens this browser's copy from then on.
 */
export async function openBackupFileAsYou(
  input: unknown,
  code: Uint8Array,
): Promise<{
  state: VaultState;
  identity: WebCryptoSharingIdentity;
  recoveryLock: ProtectedSharingIdentityV1;
}> {
  const file = parseBackupFile(input);
  const lockOnly: ProtectedSharingIdentityStore = {
    load: async (appId) => (appId === KEYWEB_RECOVERY_APP_ID ? file.recoveryLock : null),
    save: async () => undefined,
    delete: async () => undefined,
  };
  const identity = await unlockRecoveryIdentity(lockOnly, code).catch(() => {
    throw new WrongBackupCode();
  });
  if (!identity) throw new Error("This backup file has no recovery lock in it.");
  const state = await decryptSharedBackupEnvelopeV1(file.envelope, codec, identity, undefined, {
    trustedOwnerKeyId: identity.publicKey.keyId,
  });
  return { state: restorable(state), identity, recoveryLock: file.recoveryLock };
}

/** The code given does not open this file. Said plainly, not as a crypto error. */
export class WrongBackupCode extends Error {
  constructor() {
    super("That recovery code doesn't open this backup file.");
    this.name = "WrongBackupCode";
  }
}

/**
 * Put a restored vault into this browser's copy, each keyring where it lives.
 *
 * A keyring this copy already keeps in a file of its own gets its passwords in
 * that file's document, not the vault's — otherwise they would be published
 * into the index, beside keyrings they were never meant to travel with. A copy
 * with no such keyrings, which is any browser being set up from the file,
 * simply takes the lot.
 */
export async function restoreInto(
  storage: {
    readState(documentId?: string): Promise<VaultState>;
    applyRemote(incoming: VaultState, documentId?: string): Promise<VaultState>;
  },
  restored: VaultState,
): Promise<void> {
  let rest = restored;
  for (const { keyringId, datasetId } of boundDatasets(await storage.readState())) {
    const slice = extractDataset(rest, keyringId);
    if (Object.keys(slice.items).length > 0) await storage.applyRemote(slice, datasetId);
    rest = withoutDatasetItems(rest, keyringId);
  }
  await storage.applyRemote(rest);
}

function parseBackupFile(input: unknown): BackupFileV1 {
  const value = typeof input === "string" ? (JSON.parse(input) as unknown) : input;
  const file = value as Partial<BackupFileV1> | null;
  if (!file || file.format !== BACKUP_FILE_FORMAT || file.version !== 1) {
    throw new Error("That isn't a Keyweb backup file.");
  }
  if (!file.recoveryLock || !file.envelope) {
    throw new Error("That Keyweb backup file is incomplete.");
  }
  return file as BackupFileV1;
}

/**
 * A backed-up vault, ready to be merged into one that may have no Drive.
 *
 * The keyring records carry which Drive file each keyring lives in. A restore
 * is most likely to happen exactly when Drive cannot be reached, so the state
 * is handed back with every keyring's items composed in and its binding left
 * off — the passwords open here and now. When Google is back and the device
 * syncs, the index brings the bindings back with their own timestamps and
 * every keyring finds its file again.
 */
function restorable(state: VaultState): VaultState {
  const keyrings = Object.fromEntries(
    Object.entries(keyringsOf(state)).map(([id, ring]) => {
      const { dataset: _dataset, ...unbound } = ring;
      return [id, unbound];
    }),
  );
  return { ...state, keyrings: keyrings as VaultState["keyrings"] };
}
